README.md
<img align="right" src="https://user-images.githubusercontent.com/225151/75162092-3ef2f780-571d-11ea-8d4e-616ccbd1c924.png" width="200" height="200"/>

`check_prometheus_metric`
=========================
![CI](https://github.com/magenta-aps/check_prometheus_metric/workflows/CI/badge.svg)

Nagios plugin for alerting on Prometheus query results.

# Usage
```
  check_prometheus_metric.sh - Nagios plugin for checking Prometheus metrics.
  
  Usage:
    check_prometheus_metric.sh -H HOST -q QUERY -w FLOAT[:FLOAT] -c FLOAT[:FLOAT]
                               -n NAME [-m METHOD] [-O] [-i] [-p]

  Options:
    -H HOST          URL of Prometheus host to query.
    -q QUERY         Prometheus query, in single quotes, that returns a float.
    -w FLOAT[:FLOAT] Warning level value (must be a float or nagios-interval).
    -c FLOAT[:FLOAT] Critical level value (must be a float or nagios-interval).
    -n NAME          A name for the metric being checked.
    -m METHOD        Comparison method, one of gt, ge, lt, le, eq, ne.
                     (Defaults to ge unless otherwise specified).
    -C CURL_OPTS     Additional flags to curl. Can be passed multiple times. 
                     Options and option values must be passed separately.
                     e.g. -C --connect-timeout -C 10 -C --cacert -C /path/to/ca.crt
    -O               Accept NaN as an "OK" result.
    -E               Accept an empty vector (null) as an "OK" result.
    -i               Print the extra metric information into the Nagios message.
    -p               Add perfdata to check output.

  Examples:
    check_prometheus_metric -q 'up{job=\"job_name\"}' -w :1 -c :1
    # Check that job is up. If not, critical.
    
    check_prometheus_metric -q 'node_load1' -w :0.05 -c :0.1
    # Check load is below 0.05 (warning) and 0.1 (critical).

    check_prometheus_metric -q 'go_threads' -w 15:25 -c :
    # Check thread count is between 15-25, warning if outside this interval.

  Dependencies:
    Requires bash, curl, cut, echo, grep, jq and sed to be in $PATH.
```
Note: `nagios-interval` refers to [this syntax](http://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT)

# Icinga configuration
To utilize the script with [Icinga 2](https://icinga.com/docs/icinga2/), four
steps must be taken;

1. The CheckCommand definition must be added to `/etc/icinga2/conf.d/check_prometheus_metric.conf`:
```
object CheckCommand "check_prometheus_metric" {
  import "plugin-check-command"
  command = [ "/usr/lib/nagios/plugins/check_prometheus_metric" ]

  arguments = {
        "-H" = {
                value = "$check_prometheus_metric_url$"
                description = "URL of Prometheus host to query."
        }
        "-q" = {
                value = "$check_prometheus_metric_query$"
                description = "Prometheus query, that returns a float."
        }
        "-w" = {
                value = "$check_prometheus_metric_warning$"
                description = "Warning level value (float or nagios-interval)."
        }
        "-c" = {
                value = "$check_prometheus_metric_critical$"
                description = "Critical level value (float or nagios-interval)."
        }
        "-n" = {
                value = "$check_prometheus_metric_name$"
                description = "A name for the metric being checked."
        }
    }
}
```
2. The checking script, must be added to `/usr/lib/nagios/plugins/check_prometheus_metric`:
```
GIT_REPO=https://github.com/magenta-aps/check_prometheus_metric
VERSION=$(curl -sL -H "Accept: application/json" ${GIT_REPO}/releases/latest | jq -r .tag_name)
curl -sL -o . ${GIT_REPO}/releases/download/${VERSION}/check_prometheus_metric.sh
chmod +x check_prometheus_metric.sh
sudo mv check_prometheus_metric.sh /usr/lib/nagios/plugins/check_prometheus_metric
```

3. Install the required software to run the script:
```
sudo apt-get update
sudo apt-get install -y bash coreutils curl grep jq sed
```

4. Add Service definitions for whatever is to be monitored:

`/etc/icinga2/conf.d/check_pi_service.conf`:
```
apply Service "pi" {
  import "generic-service"

  check_command = "check_prometheus_metric"

  vars.check_prometheus_metric_url = "prometheus:9090"
  vars.check_prometheus_metric_query = "pi"
  vars.check_prometheus_metric_warning = "3:4"
  vars.check_prometheus_metric_critical = "1:6"
  vars.check_prometheus_metric_name = "pi"
  
  command_endpoint = host.vars.client_endpoint
  assign where host.name == NodeName
}
```

# Nagios configuration
You need to add the following commands to your Nagios configuration to use it:
```
define command {
    command_name check_prometheus
    command_line $USER1$/check_prometheus_metric.sh -H '$ARG1$' -q '$ARG2$' -w '$ARG3$' -c '$ARG4$' -n '$ARG5$' -m '$ARG6$'
}
```


CHANGELOG
0.0.5
=====
Fixed trailing whitespace (#2)
Fixed a bug with not passing curl context on all calls (#3)

0.0.4
=====
Added CHANGELOG
Added logo in README.md
Added configuration documentation to README.md
Added additional preflight check for validation prometheus server connection.
Added test case against icinga itself.
Fixed a typo in the README.md file
Fixed bug causing wrong exit code for WARNING/CRITICAL returns

0.0.3
=====
Rework plugin to new structure.

Added nagios-interval syntax support
Added bats testing framework
- Added tests for current (old) behavior and syntax.
- Added tests for new behavior and syntax.
Autodetect vector / scalar query types based on query result.
Check intervals using jq instead of test -v
Support floats instead of integers all around.
Utilize epsilon instead of 1 for old syntax.
Added CI testing and release management.

0.0.2
=====
Rework project structure

0.0.1
=====
Created release workflow


test.sh
#!/bin/bash
NETWORK_NAME=nagios_plugins_testnetwork
PROMETHEUS_NAME=nagios_plugins_prometheus
PROMETHEUS_PORT=9090
PUSHGATEWAY_NAME=nagios_plugins_pushgateway
PUSHGATEWAY_PORT=9091
ICINGA_NAME=icinga2
ICINGA_PORT=5665
NGINX_NAME=nagios_plugins_nginx
NGINX_PORT=8022

PLUGIN_SCRIPT=build/output.sh

export PLUGIN_SCRIPT=${PLUGIN_SCRIPT}
export PROMETHEUS_PORT=${PROMETHEUS_PORT}
export PUSHGATEWAY_PORT=${PUSHGATEWAY_PORT}
export ICINGA_PORT=${ICINGA_PORT}
export NGINX_PORT=${NGINX_PORT}

source tests/test_utils.bash

build_script

start_docker_network
start_prometheus
start_pushgateway
start_icinga
start_nginx_basic_auth

set_metric "pi" "3.14"

echo ""
echo "Waiting until prometheus sees pushed metric"
wait_for_metric "scalar(pi)" 3.14
echo ""

echo "Ordinary tests"
bats tests/*.bats
ORDINARY_TESTS_STATUS=$?
echo "Ordinary tests status: ${ORDINARY_TESTS_STATUS}"

echo "Integration tests"
bats tests/icinga/*.bats
INTEGRATION_TESTS_STATUS=$?
echo "Integration tests status: ${INTEGRATION_TESTS_STATUS}"

! (( ${ORDINARY_TESTS_STATUS} || ${INTEGRATION_TESTS_STATUS} ))


tests/utils.sh
function wait_for_metric() {
    PARSED_RESULT="NaN"
    # Keep running query until data is found, i.e. non-nan returned
    until [ "${PARSED_RESULT}" != "NaN" ]; do
        printf '.'
        sleep 1
        RAW_RESULT=$(curl -s --data-urlencode "query=$1" "http://localhost:$2/api/v1/query")
        PARSED_RESULT=$(echo "${RAW_RESULT}" | jq -r .data.result[1])
    done
}

TOTAL_TESTS=0
TOTAL_FAILS=0
function check() {
    # TODO: Check exit-code against OK/WARNING/CRITICAL/UNKNOWN
    TOTAL_TESTS=$((TOTAL_TESTS + 1))
    if [ "$1" != "$2" ]; then
        TOTAL_FAILS=$((TOTAL_FAILS + 1))
        printf "x"
        printf "\n$1 != $2 ($3)\n"
    else
        printf "."
    fi
}

function test_line() {
    # Loop through each test-case
    while IFS= read -r line; do
        PARAMETERS=$(echo "$line" | cut -f1 -d'#')
        EXPECTED=$(echo "$line" | cut -f2 -d'#')
        OUTPUT=$(bash ${PLUGIN_SCRIPT} -H "${PROMETHEUS_SERVER}" ${PARAMETERS} -n "tc")
        EXIT_CODE=$?
        RESULT=$(echo "${OUTPUT}" | head -1)
        check "$RESULT" "$EXPECTED" "$line" "$EXIT_CODE"
    done <<< "$1"
}



function parameterized_tests() {
    # Loop through each test-case
    while IFS= read -r line; do
        PARAMETERS=$(echo "$line" | cut -f1 -d'#')
        EXPECTED=$(echo "$line" | cut -f2 -d'#')
        OUTPUT=$(bash ${PLUGIN_SCRIPT} -H "${PROMETHEUS_SERVER}" ${PARAMETERS} -n "tc")
        EXIT_CODE=$?
        RESULT=$(echo "${OUTPUT}" | head -1)
        check "$RESULT" "$EXPECTED" "$line" "$EXIT_CODE"
    done <<< "$1"
}

function print_exit() {
    $1
    echo $?
}


tests/test_nan.sh
#!/bin/bash
set +x

PLUGIN_SCRIPT=$1
PROMETHEUS_PORT=$2
PUSHGATEWAY_PORT=$3

PROMETHEUS_SERVER=http://localhost:${PROMETHEUS_PORT}

source tests/utils.sh

# Parameterized test-cases, format:
# commandline-parameters #expected-output
TESTS="
# Base-case
-q 1 -w 2 -c 3 #OK - tc is 1
-q 1 -w 2 -c 3 -O #OK - tc is 1

# NaN as return
-q NaN -w 2 -c 3 #UNKNOWN - unable to parse prometheus response
-q NaN -w 2 -c 3 -O #OK - tc is NaN
"
# Get rid of empty lines and comments
TESTS=$(echo "${TESTS}" | sed '/^$/d' | sed '/^#/d')
# Loop through each test-case
parameterized_tests "${TESTS}"

echo ""
echo "$((TOTAL_TESTS - TOTAL_FAILS)) / ${TOTAL_TESTS}"
exit ${TOTAL_FAILS}


src/parse.bash
FLOAT_REGEX="[+-]?([0-9]*[.])?[0-9]+"
INTERVAL_REGEX="([~]|${FLOAT_REGEX}):(${FLOAT_REGEX})?"

function is_inverted {
    # Inverted if string starts with @
    local _IS_INVERTED
    echo "${1}" | grep -E "^@" -c >/dev/null
    _IS_INVERTED=$?
    return ${_IS_INVERTED}
}

function is_float() {
    local _IS_FLOAT
    echo "${1}" | grep -E "^${FLOAT_REGEX}$" -c >/dev/null
    _IS_FLOAT=$?
    return ${_IS_FLOAT}
}

function is_interval() {
    local _IS_INTERVAL=0
    echo "${1}" | grep -E "^${INTERVAL_REGEX}$" -c >/dev/null
    _IS_INTERVAL=$?
    return ${_IS_INTERVAL}
}

function is_float_or_interval() {
    if is_float "${1}" || is_interval "${1}"; then
        return 0
    fi
    return 1
}

function decode_range() {
    # Decode Nagios Threshold format string.
    #
    # For reference, see: https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT
    #
    # Examples:
    #   Input: "10:20", Output: "10 20"
    #   Input: "@~:3.14", Output: "-inf 3.14 inverted"
    # Input variable
    local _INPUT=$1
    # Output variables
    local _START="0"
    local _END="inf"
    local _INVERTED="0"
    # Check if inverted
    if is_inverted "${_INPUT}"; then
        _INVERTED="1"
        # Remove @ from string
        _INPUT="${_INPUT:1}"
    fi
    # Check if lonely float (i.e. implicit interval)
    if is_float "${_INPUT}"; then
        _END=${_INPUT}
    # Check if valid interval
    elif is_interval "${_INPUT}"; then
        # Fetch parts of interval seperately
        _START=$(echo "${_INPUT}" | cut -f1 -d':')
        _END=$(echo "${_INPUT}" | cut -f2 -d':')
        # Replace ~ in start with -inf
        _START=$(echo "${_START}" | sed "s/^~$/-inf/g")
        # Replace empty in end with inf
        _END=$(echo "${_END}" | sed "s/^$/inf/g")
    else  # Not valid float or interval
        echo "Unable to parse range"
        return 1
    fi
    # Strip prefix +
    _START=$(echo ${_START} | cut -f2 -d'+')
    _END=$(echo ${_END} | cut -f2 -d'+')
    # Output space-seperated string
    printf '%s' "${_START} ${_END} ${_INVERTED}"
    return 0
}


src/check_prometheus_metric.sh
#!/bin/bash
#
# check_prometheus_metric.sh - Nagios plugin wrapper for checking Prometheus
#                              metrics. Requires curl and jq to be in $PATH.

# Default configuration:
CURL_OPTS=()
COMPARISON_METHOD=ge
NAN_OK="false"
NAGIOS_INFO="false"
PERFDATA="false"
PROMETHEUS_QUERY_TYPE=""

# Constants
#----------
# Nagios status codes:
OK=0
WARNING=1
CRITICAL=2
UNKNOWN=3

# Variables:
NAGIOS_STATUS=UNKNOWN
NAGIOS_SHORT_TEXT='an unknown error occured'
NAGIOS_LONG_TEXT=''

# Includes
source preamble.bash
source parse.bash
source usage.bash

# Main

# Code:
function check_dependencies() {
    if ! [ -x "$(command -v curl)" ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT='missing "curl" command'
        exit
    fi

    if ! [ -x "$(command -v jq)" ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT='missing "jq" command'
    fi
}

function process_command_line {

  while getopts ':H:q:w:c:m:n:C:Oipt:' OPT "$@"
  do
    case ${OPT} in
      H)        PROMETHEUS_SERVER="$OPTARG" ;;
      q)        PROMETHEUS_QUERY="$OPTARG" ;;
      n)        METRIC_NAME="$OPTARG" ;;

      m)        # If invalid operator name
                if ! [[ ${OPTARG} =~ ^([lg][et]|eq|ne)$ ]]; then
                    NAGIOS_SHORT_TEXT="invalid comparison method: ${OPTARG}"
                    NAGIOS_LONG_TEXT="$(usage)"
                    exit
                fi
                COMPARISON_METHOD=${OPTARG}
                ;;

      c)        # If malformed
                if ! is_float_or_interval "${OPTARG}"; then
                  NAGIOS_SHORT_TEXT='-c CRITICAL_LEVEL requires a float or interval'
                  NAGIOS_LONG_TEXT="$(usage)"
                  exit
                fi
                CRITICAL_LEVEL=${OPTARG}
                ;;

      w)        # If malformed
                if ! is_float_or_interval "${OPTARG}"; then
                  NAGIOS_SHORT_TEXT='-w WARNING_LEVEL requires a float or interval'
                  NAGIOS_LONG_TEXT="$(usage)"
                  exit
                fi
                WARNING_LEVEL=${OPTARG}
                ;;

      C)        CURL_OPTS+=("${OPTARG}")
                ;;
      O)        NAN_OK="true"
                ;;

      i)        NAGIOS_INFO="true"
                ;;

      p)        PERFDATA="true"
                ;;

      t)        
                NAGIOS_LONG_TEXT+="Note: The use of -t is deprecated, as the query-type is derived from the query result."
                ;;

      \?)       NAGIOS_SHORT_TEXT="invalid option: -$OPTARG"
                NAGIOS_LONG_TEXT="$(usage)"
                exit
                ;;

      \:)       NAGIOS_SHORT_TEXT="-$OPTARG requires an arguement"
                NAGIOS_LONG_TEXT="$(usage)"
                exit
                ;;
    esac
  done

  # check for missing parameters
  if [[ -z ${PROMETHEUS_SERVER} ]] ||
     [[ -z ${PROMETHEUS_QUERY} ]] ||
     [[ -z ${METRIC_NAME} ]] ||
     [[ -z ${WARNING_LEVEL} ]] ||
     [[ -z ${CRITICAL_LEVEL} ]]
  then
    NAGIOS_SHORT_TEXT='missing required option'
    NAGIOS_LONG_TEXT+="$(usage)"
    exit
  fi
  # Convert old syntax to new syntax
  # Warning
  if is_float "${WARNING_LEVEL}"; then
      if [[ "${COMPARISON_METHOD}" == "ge" ]]; then
          WARNING_LEVEL="0:$(echo ${WARNING_LEVEL} | jq '. - 1e-9')"
      elif [[ "${COMPARISON_METHOD}" == "eq" ]]; then
          WARNING_LEVEL="@${WARNING_LEVEL}:${WARNING_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "ne" ]]; then
          WARNING_LEVEL="${WARNING_LEVEL}:${WARNING_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "lt" ]]; then
          WARNING_LEVEL="${WARNING_LEVEL}:"
      elif [[ "${COMPARISON_METHOD}" == "le" ]]; then
          WARNING_LEVEL="$(echo ${WARNING_LEVEL} | jq '. + 1e-9'):"
      fi
  fi
  # Critical
  if is_float "${CRITICAL_LEVEL}"; then
      if [[ "${COMPARISON_METHOD}" == "ge" ]]; then
          CRITICAL_LEVEL="0:$(echo ${CRITICAL_LEVEL} | jq '. - 1e-9')"
      elif [[ "${COMPARISON_METHOD}" == "eq" ]]; then
          CRITICAL_LEVEL="@${CRITICAL_LEVEL}:${CRITICAL_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "ne" ]]; then
          CRITICAL_LEVEL="${CRITICAL_LEVEL}:${CRITICAL_LEVEL}"
      elif [[ "${COMPARISON_METHOD}" == "lt" ]]; then
          CRITICAL_LEVEL="${CRITICAL_LEVEL}:"
      elif [[ "${COMPARISON_METHOD}" == "le" ]]; then
          CRITICAL_LEVEL="$(echo ${CRITICAL_LEVEL} | jq '. + 1e-9'):"
      fi
  fi
  # Decode new syntax
  # Warning
  WARNING_RANGE=$(decode_range ${WARNING_LEVEL})
  WARNING_LEVEL_LOW=$(echo "${WARNING_RANGE}" | cut -f1 -d' ')
  WARNING_LEVEL_HIGH=$(echo "${WARNING_RANGE}" | cut -f2 -d' ')
  WARNING_INVERTED=$(echo "${WARNING_RANGE}" | cut -f3 -d' ')
  # Critical
  CRITICAL_RANGE=$(decode_range ${CRITICAL_LEVEL})
  CRITICAL_LEVEL_LOW=$(echo "${CRITICAL_RANGE}" | cut -f1 -d' ')
  CRITICAL_LEVEL_HIGH=$(echo "${CRITICAL_RANGE}" | cut -f2 -d' ')
  CRITICAL_INVERTED=$(echo "${CRITICAL_RANGE}" | cut -f3 -d' ')
  # Bake our levels json
  LEVEL_JSON="{\"critical_low\": ${CRITICAL_LEVEL_LOW}, \"critical_high\": ${CRITICAL_LEVEL_HIGH}, \"warning_low\": ${WARNING_LEVEL_LOW}, \"warning_high\": ${WARNING_LEVEL_HIGH}}"
  # Sanity check critical and warning levels
  if [ ${COMPARISON_METHOD} != "ne" ]; then
      echo "${LEVEL_JSON}" | jq -e ".critical_low <= .critical_high" >/dev/null
      CRITICAL_LEVEL_OK=$?
      echo "${LEVEL_JSON}" | jq -e ".warning_low <= .warning_high" >/dev/null
      WARNING_LEVEL_OK=$?
      if [ ${CRITICAL_LEVEL_OK} -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="invalid critical range: ${CRITICAL_LEVEL_LOW} <= ${CRITICAL_LEVEL_HIGH}"
        exit
      fi
      if [ ${WARNING_LEVEL_OK} -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="invalid warning range: ${WARNING_LEVEL_LOW} <= ${WARNING_LEVEL_HIGH}"
        exit
      fi
  fi
}


function check_prometheus_server {
    # Check that the server URL responds
    curl --silent --head ${PROMETHEUS_SERVER}/api/v1/query > /dev/null
    SERVER_RESPONDED=$?
    if [ "${SERVER_RESPONDED}" -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="no response from prometheus"
        exit
    fi

    # Check if the prometheus endpoint exists, and can be queried
    curl -s "${CURL_OPTS[@]}" ${PROMETHEUS_SERVER}/api/v1/query?query=cafebabe | jq -e '.status == "success"' 1>/dev/null 2>/dev/null
    PROMETHEUS_OK=$?
    if [ "${PROMETHEUS_OK}" -ne 0 ]; then
        NAGIOS_STATUS=UNKNOWN
        NAGIOS_SHORT_TEXT="unable to query prometheus endpoint!"
        exit
    fi
}


function get_prometheus_raw_result {

  local _RESULT

  _RESULT=$(curl -sgG "${CURL_OPTS[@]}" --data-urlencode "query=${PROMETHEUS_QUERY}" "${PROMETHEUS_SERVER}/api/v1/query")
  printf '%s' "${_RESULT}"

}

function get_prometheus_scalar_result {

  local _RESULT

  _RESULT=$(echo $1 | jq -r '.[1]')

  # check result
  case "${_RESULT}" in
    +Inf) printf '%s' 'inf'
          ;;
    -Inf) printf '%s' '-inf' 
          ;;
    *)    printf '%s' "${_RESULT}" # otherwise return as a string
          ;;
  esac
}

function get_prometheus_vector_value {

  local _RESULT

  # return the value of the first element of the vector
  _RESULT=$(echo $1 | jq -r '.[0].value?')
  printf '%s' "${_RESULT}"

}

function get_prometheus_vector_metric {

  local _RESULT

  # return the metric information of the first element of the vector
  _RESULT=$(echo $1 | jq -r '.[0].metric?' | xargs)
  printf '%s' "${_RESULT}"

}

if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
    # Exit trigger
    #-------------
    # Function to be trapped on exit
    function on_exit {
        if [[ -n ${NAGIOS_LONG_TEXT} ]]; then
            printf '%s\n' "${NAGIOS_LONG_TEXT}"
        fi
        printf '%s - %s\n' ${NAGIOS_STATUS} "${NAGIOS_SHORT_TEXT}"
        # Indirect variable reference
        exit ${!NAGIOS_STATUS}
    }

    # Set up exit function
    trap on_exit EXIT TERM

    check_dependencies
    # process the cli options
    process_command_line "$@"

    # Check that we can communicate with the prometheus server
    check_prometheus_server

    # get the raw query from prometheus
    PROMETHEUS_RAW_RESPONSE="$( get_prometheus_raw_result )"

    PROMETHEUS_QUERY_TYPE=$(echo "${PROMETHEUS_RAW_RESPONSE}" | jq -r '.data.resultType')
    PROMETHEUS_RAW_RESULT=$(echo "${PROMETHEUS_RAW_RESPONSE}" | jq -r '.data.result')

    # extract the metric value from the raw prometheus result
    if [[ "${PROMETHEUS_QUERY_TYPE}" == "scalar" ]]; then
        PROMETHEUS_RESULT=$( get_prometheus_scalar_result "$PROMETHEUS_RAW_RESULT" )
        PROMETHEUS_METRIC=UNKNOWN
    else
        PROMETHEUS_VALUE=$( get_prometheus_vector_value "$PROMETHEUS_RAW_RESULT" )
        PROMETHEUS_RESULT=$( get_prometheus_scalar_result "$PROMETHEUS_VALUE" )
        PROMETHEUS_METRIC=$( get_prometheus_vector_metric "$PROMETHEUS_RAW_RESULT" ) 
    fi

    # check the value
    if is_float ${PROMETHEUS_RESULT}; then
      JSON=$(echo "${LEVEL_JSON} {\"value\": ${PROMETHEUS_RESULT}}" | jq -s add)
      # Evaluate critical and warning levels
      echo "${JSON}" | jq -e ".critical_low <= .value and .value <= .critical_high" >/dev/null
      IS_CRITICAL=$?
      echo "${JSON}" | jq -e ".warning_low <= .value and .value <= .warning_high" >/dev/null
      IS_WARNING=$?

      if [ ${IS_CRITICAL} -ne ${CRITICAL_INVERTED} ]; then
        NAGIOS_STATUS=CRITICAL
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      elif [ ${IS_WARNING} -ne ${WARNING_INVERTED} ]; then
        NAGIOS_STATUS=WARNING
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      else
        NAGIOS_STATUS=OK
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      fi
    else
      if [[ "${NAN_OK}" = "true" && "${PROMETHEUS_RESULT}" = "NaN" ]]; then
        NAGIOS_STATUS=OK
        NAGIOS_SHORT_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      else    
        NAGIOS_SHORT_TEXT="unable to parse prometheus response"
        NAGIOS_LONG_TEXT="${METRIC_NAME} is ${PROMETHEUS_RESULT}"
      fi
    fi
    if [[ "${NAGIOS_INFO}" = "true" ]]; then
        NAGIOS_SHORT_TEXT="${NAGIOS_SHORT_TEXT}: ${PROMETHEUS_METRIC}"
    fi
    if [[ "${PERFDATA}" = "true" ]]; then
        # Bake performance data
        PERF_DATA=""
        PERF_DATA+="query_result=${PROMETHEUS_RESULT};${WARNING_LEVEL};${CRITICAL_LEVEL};U;U"
        # PERF_DATA+=" "

        NAGIOS_SHORT_TEXT="${NAGIOS_SHORT_TEXT} | ${PERF_DATA}"
    fi

    exit
fi


src/usage.bash
function usage() {

  cat <<'EoL'

  check_prometheus_metric.sh - Nagios plugin wrapper for checking Prometheus
                               metrics. Requires curl and jq to be in $PATH.

  Usage:
  check_prometheus_metric.sh -H HOST -q QUERY -w FLOAT[:FLOAT] -c FLOAT[:FLOAT] -n NAME [-m METHOD] [-O] [-i] [-p] [-t QUERY_TYPE]

  Options:
    -H HOST          URL of Prometheus host to query.
    -q QUERY         Prometheus query, in single quotes, that returns by default a float or int (see -t).
    -w FLOAT[:FLOAT] Warning level value (must be a float or nagios-interval).
    -c FLOAT[:FLOAT] Critical level value (must be a float or nagios-interval).
    -n NAME          A name for the metric being checked.
    -m METHOD        Comparison method, one of gt, ge, lt, le, eq, ne.
                     (Defaults to ge unless otherwise specified.)
    -C CURL_OPTS     Additional flags to pass to curl.
                     Can be passed multiple times. Options and option values must be passed separately.
                     e.g. -C --conect-timetout -C 10 -C --cacert -C /path/to/ca.crt
    -O               Accept NaN as an "OK" result .
    -i               Print the extra metric information into the Nagios message.
    -p               Add perfdata to check output.

  Examples:
    check_prometheus_metric -q 'up{job=\"job_name\"}' :1 -c :1  # Check that job is up.
    
    check_prometheus_metric -q 'node_load1' -w :0.05 -c :0.1  # Check load is OK.
    # Aka. that load is below 0.05 and 0.1 respectively.

    check_prometheus_metric -q 'go_threads' -w 15:25 -c :  # Check thread count is OK.
    # Aka. OK if we have 15-25 threads, outside of this; warning, never critical.

EoL
}
